source/extensions/filters/http/cache/cache_header_utility.cc

```diff
@@ -0,0 +1,107 @@
+#include "extensions/filters/http/cache/cache_header_utility.h"
+
+#include <vector>
+
+#include "common/common/utility.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/header_utility.h"
+#include "common/http/headers.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/strip.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+namespace {
+const absl::string_view bytes_{"bytes"};
+}
+
+std::vector<RawByteRange> CacheHeaderUtility::getRanges(const Http::HeaderMap& request_headers) {
+  if (request_headers.Method() == nullptr ||
+      request_headers.Method()->value().getStringView() != Http::Headers::get().MethodValues.Get) {
+    // Range headers are only valid on GET requests so don't bother parsing
+    // range if we're not going to use it.
+    return {};
+  }
+
+  // Multiple instances of range/range-unit headers are considered invalid.
+  // https://tools.ietf.org/html/rfc7230#section-3.2.2
+  // TODO: should we log when encountering multiple headers?
+  std::vector<absl::string_view> range_unit_headers;
+  Http::HeaderUtility::getAllOfHeader(request_headers, Http::Headers::get().RangeUnit.get(),
+                                      range_unit_headers);
+  absl::string_view range_unit;
+  if (range_unit_headers.size() == 1) {
+    range_unit = range_unit_headers.front();
+  } else if (range_unit_headers.empty()) {
+    range_unit = bytes_;
+  } else {
+    return {};
+  }
+
+  std::vector<absl::string_view> range_headers;
+  Http::HeaderUtility::getAllOfHeader(request_headers, Http::Headers::get().Range.get(),
+                                      range_headers);
+  absl::string_view range;
+  if (range_headers.size() == 1) {
+    range = range_headers.front();
+  } else {
+    return {};
+  }
+
+  return parseRangeHeaderValue(range_unit, range);
+}
+
+std::vector<RawByteRange> CacheHeaderUtility::parseRangeHeaderValue(absl::string_view range_unit,
+                                                               absl::string_view range) {
+  if (range_unit.empty() || !absl::ConsumePrefix(&range, range_unit) ||
+      !absl::ConsumePrefix(&range, "=")) {
+    return {};
+  }
+
+  std::vector<RawByteRange> ranges;
+  while (!range.empty()) {
+    uint64_t first, last;
+    if (!StringUtil::consumeLeadingDigits(&range, &first)) {
+      if (!absl::ConsumePrefix(&range, "-")) {
+        ranges.clear();
+        break;
+      }
+      first = UINT64_MAX;
+    }
+
+    if (!absl::ConsumePrefix(&range, "-")) {
+      if (!StringUtil::consumeLeadingDigits(&range, &first)) {
+        ranges.clear();
+        break;
+      }
+    }
+
+    if (!StringUtil::consumeLeadingDigits(&range, &last)) {
+      if (!range.empty()) {
+        ranges.clear();
+        break;
+      }
+      last = first;
+      first = UINT64_MAX;
+    }
+
+    ranges.push_back(RawByteRange(first, last));
```

Comment:
possibly have processing to sniff out out of order ranges or overlapping ranges as RFC states that `both are indications of either a broken client or a deliberate denial-of-service attack`?

Reply:
I think we should just have a configurable limit on how many ranges we allow, defaulting to 0.

Reply:
I renamed the function to `readAndRemoveLeadingDigits`, made it take a reference and return `optional`, and renamed `v`.

I also unified the int parsing in both eatLeadingDuration and getRanges in the latest commit and the code isn't awful, tho it isn't super clean either. I'm happy to revert it to the old way, but I figured I'd get some input before doing that.

Reply:
@toddmgreer I'm looking into how config works in Envoy. Do you think this PR should wait till I add cache config functionality to it? It's fine by me either way, just curious what your thoughts are.

Reply:
I added a first swag at using config to control this limit, but I'm not sure I'm pleased with how it came out. What do you think? I am unsure where we should store the config values and if maybe we should pass a config object to the LookupRequest constructor instead of the byte range limit value.